src/coreclr/jit/hwintrinsiccodegenarm64.cpp

```diff
@@ -824,7 +824,30 @@ void CodeGen::genHWIntrinsic(GenTreeHWIntrinsic* node)
                         assert(!node->IsEmbMaskOp());
                         if (HWIntrinsicInfo::IsExplicitMaskedOperation(intrin.id))
                         {
-                            GetEmitter()->emitIns_R_R_R(ins, emitSize, targetReg, op1Reg, op2Reg, opt);
+                            if (isRMW)
+                            {
+                                if (targetReg != op2Reg)
+                                {
+                                    assert(targetReg != op1Reg);
+
+                                    instruction insMov = INS_mov;
+
+                                    if (GetEmitter()->isPredicateRegister(targetReg) &&
```

Comment:
can we just assert here and just set to `INS_sve_mov` because we are already in `IsExplicitMaskedOperation()`?

```c++
assert(GetEmitter()->isPredicateRegister(targetReg) && GetEmitter()->isPredicateRegister(op2Reg))
```

Reply:
Sure

Reply:
Such asserts can hurt later refactorings, we did quite a bit to explicitly centralize such lookups and reduce the number of direct instruction usages because it made it easier to light up new ISAs and insert broader assertions and optimizations

Reply:
This in general should probably be using the general ins_mov helper and possibly the emitIns_move helper to ensure such optimizations also get picked up here

Reply:
> ins_mov helper 

Yes, that is the plan. Its just that it won't happen in .NET 9

> Such asserts can hurt later refactorings

The reason I asked is so we are conservative and know what scenarios the code will handle. We do not want something the code get executed for scenario that was meant to handle it properly.

Reply:
Why wouldn't moving of predicate registers be added to the `ins_mov` helper? They are a unique type that will only be encountered for SVE code

Reply:
> Why wouldn't moving of predicate registers be added to the `ins_mov` helper? They are a unique type that will only be encountered for SVE code

Are you referring to `ins_Mov_Extend()` method? Seems we already have `INS_sve_mov` behind `unreached()`. I think I tried removing it at one point, but ran into asserts or something that needed bigger changes. Perhaps at that time the predicate registers were not separated. But if it is as simple as deleting the `unreached()` and then using `ins_Mov_Extend()` then I am ok with it. If it needs bigger changes, lets defer it ATM.

Reply:
https://github.com/dotnet/runtime/blob/851100269d7f286195d7a14cf3ac47584fe59c68/src/coreclr/jit/instr.cpp#L1723-L1726

Reply:
I expect we require this working for the broader support of spill/restore and shuffling of data to work in .NET 9.

I would expect it is in the same general bucket as callee save/trash and thread suspension support, where we will have subtle bugs/crashes if it isn't there.

Reply:
> ins_Mov_extend

After looking through this, I think this was just missed in https://github.com/dotnet/runtime/pull/101885, where we eliminated `unreached()` in `instr.cpp`. 

> possibly the emitIns_move helper to ensure such optimizations also get picked up here

We fixed this helper in #101885, #102218 and #102007 to handle all the cases

@TIHan - I think you should remove the remaining `unreached()` from `instr.cpp` too and then should be able to use `ins_Mov_extend()` helper.

Reply:
Will do.